### ISSUE TITLE
Get the server run in development mode but make it thinks is building

### ISSUE THREAD

**Issue by andreamoro** [middleman/middleman#2311]
Apologies for the tricky title, but I guess it explains everything? I have also started a [thread](https://forum.middlemanapp.com/t/cli-run-middleman-local-server-but-making-it-believes-is-in-building-mode/4336) in the middleman forum, but it doesn't seem to be quite active these days.

## Expected behavior and actual behavior
I need to run middelman with the attached server but let it pretending it's building the site. This is because I need to test what the URLs and some other bits will look like by the time the site is built, but without waiting for the whole set of pages being processed.

I found something similar to what I'm asking, if not exactly that, mentioned in this [issue](https://github.com/middleman/middleman/pull/1293). 

Regrettably, by adding a puts app.production? in within the configure: development block returns false thus suggesting this is no longer supported.

## Steps to reproduce the problem (from a clean middleman installation)

1) In config.rb within the `configure: development` add the following puts
```
  puts app.development? 
  puts app.build? 
  puts app.production?
```
2) Run a `bundle exec middleman server -e production` (which seems still [supported](https://www.rubydoc.info/github/middleman/middleman/Middleman%2FApplication:production%3F))

3) check the output, which in my case is not what expected

```
[1] $ bundle exec middleman server -e production
true
false
false
```

## Additional information

- Ruby version: 2.5.0p0
- Middleman version: 4.2
- OS version: Mojave

**Comment by andreamoro** [comment]
It actually turned out that the parameter works, but the `app.production?` variable doesn't seem to return the correct value which is somehow expected.

Looking at the middleman/config when the server is running, I can clearly see the following when a `-e production` is appended at CLI.

```
:cli_options = {"verbose"=>false,
 "instrument"=>false,
 "profile"=>false,
 "daemon"=>false,
 "environment"=>"production"}
```

**Comment by andreamoro** [comment]
Ultimately, not sure this [issue](https://github.com/middleman/middleman/issues/1748) is related, but if in the config.rb I add something like

```
configure :production do
  puts 'Hello'
end
```

The configuration section is ignored. 

@tdreyno can you please shed a light?

**Comment by tdreyno** [comment]
I think you're right, the env in dev mode is getting overwritten.

In the short time, can you try:

```
MM_ENV=production bundle exec middleman server
````

I'll dig in and find a real fix.

**Comment by andreamoro** [comment]
OK, so I understand what happens now. 

By appending a -e <something> to bundle exec middleman server, the environment is set immediately after the development one is run, which I'm not sure it is how was originally intended?

Take the code below.

```
configure :production do
  puts 'app.production?: ' << app.production?.to_s
  puts 'app.development?: ' << app.development?.to_s 

  set :protocol, "https://"
  set :domain, "www.production.me"
  set :port, 80  

  puts config
end

configure :development do
  puts 'app.production?: ' << app.production?.to_s
  puts 'app.development?: ' << app.development?.to_s 

  set :protocol, "https://"
  set :domain, "www.site.dev"
  set :port, 8081
  set :css_dir, "css"

  puts config
end
```

The output will we:

```
app.production?: false
app.development?: true

.... config variables with 

- domain set as www.site.dev 
- port 8081
- css_dir to css

app.production?: true
app.development?: false

.... config variables with 

- domain set as www.production.me 
- port 80
- css_dir to stylesheets

```

The same behaviour occurs with any other environment variable, which exclude the possibility `production` being treated differently.

So basically the production environment block is executed, but immediately after the development one. Also any default variable like the `css_dir` is restored to the default value when the new configuration block is executed. Something questionable I'd say (I can see both pro and cons in variable staying or being restored).

In an ideal scenario, the following should happen:
- only one environment block should be executed, which is the case of the MM_ENV being declared as you previously suggested.
- a cli_object variable should be made immediately available and independently of the config

**Comment by tdreyno** [comment]
Can you try the above env variable approach? I would expect 2 initialization still, but of the correct mode.

Basically, the CLI boots up before the server, which is why there are 2 initializations. The CLI boot is so that some of the other commands know config values from config.rb, but probably shouldn't be necessary for `server` and `build`.

**Comment by andreamoro** [comment]
Sorry, I thought I did answer already.
By using the variable approach I can see only one initialization.

I will add back some debug line and print out the results later.

**Comment by andreamoro** [comment]
I may be wrong, again because I have not investigated into the source code, but could this double initialisation due to the blog module?

![image](https://user-images.githubusercontent.com/5256583/73718827-dd42fd00-471d-11ea-83c4-b1bc26e7ce1f.png)

If you see the screenshot above, the `true` just before `The middleman is loading` is set in a `configure :<something> do` block, but it is read after the blog module is initialised. 

Quite annoying as the sole purpose of this thread was to change that settings  based on the type of environment I'm on.

**Comment by tdreyno** [comment]
No, the double initialization is because the cli boots once to detect valid flags. Then the server boots after.
